File: 2191-sort-the-jumbled-numbers/2191-sort-the-jumbled-numbers.java
class Trio implements Comparable<Trio> {
    int val;
    int ref;
    int idx;

    public Trio(int val, int ref, int idx) {
        this.val = val;
        this.ref = ref;
        this.idx = idx;
    }

    public int compareTo(Trio o) {
        if (this.ref < o.ref) {
            return -1;
        } else if (this.ref > o.ref) {
            return 1;
        } else if (this.idx < o.idx) {
            return -1;
        } else if (this.idx > o.idx) {
            return 1;
        }
        return 0;
    }
}

class Solution {
    private int getVal(int n, int[] mapping) {
        String target = Integer.toString(n);
        int res = 0;

        for (int i = 0; i < target.length(); i++) {
            res = res * 10 + mapping[target.charAt(i) - '0'];
        }

        return res;
    }

    public int[] sortJumbled(int[] mapping, int[] nums) {
        Trio[] temp = new Trio[nums.length];
        int[] result = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            temp[i] = new Trio(nums[i], getVal(nums[i], mapping), i);
        }

        Arrays.sort(temp);

        for (int i = 0; i < nums.length; i++) {
            result[i] = temp[i].val;
        }

        return result;
    }
}

File: 2191-sort-the-jumbled-numbers/2191-sort-the-jumbled-numbers.py
class Solution:
    def sortJumbled(self, mapping: List[int], nums: List[int]) -> List[int]:
        def get_val(n):
            target = str(n)
            res = 0
            for i in range(len(target)):
                res = res * 10 + mapping[int(target[i])]
            return res


        result = list()

        for i in range(len(nums)):
            result.append((nums[i], get_val(nums[i]), i))

        return [x[0] for x in sorted(result, key=lambda x: (x[1], x[2]))]
